### PR TITLE
Update JsonSchema.Net version

### PR DESCRIPTION
**Pull request by gregsdennis**
# PR Summary

Follow-up to #18141.  Updates JsonSchema.Net to latest version: v3.3.2 -> v4.1.0.

Release notes can be viewed [here](https://json-everything.net/json-schema#release-schema-4.1.0).

## PR Context

While not strictly required now, this PR sets up for later updates that will incorporate additional support for JSON Schema.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/resources/TestJsonCmdletStrings.resx [hunk 1]
```diff
@@ -124,7 +124,7 @@
     <value>Cannot parse the JSON.</value>
   </data>
   <data name="InvalidJsonAgainstSchemaDetailed" xml:space="preserve">
-    <value>The JSON is not valid with the schema: {0} at {1}</value>
```
Comment (gregsdennis) [inline src/Microsoft.PowerShell.Commands.Utility/resources/TestJsonCmdletStrings.resx:127]: When I updated, I tested this locally with a schema & instance that had a problem at the root.  The error that's output just ends with "at", and the fact that there's an empty JSON Pointer indicating the root object isn't conveyed.  I think the quotes help.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs [hunk 2]
```diff
@@ -239,29 +239,17 @@ protected override void ProcessRecord()
 
                 if (_jschema != null)
                 {
-                        var validationResults = _jschema.Validate(parsedJson, new ValidationOptions { OutputFormat = OutputFormat.Basic });
-                        result = validationResults.IsValid;
+                        var evaluationResults = _jschema.Evaluate(parsedJson, new EvaluationOptions { OutputFormat = OutputFormat.List });
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs:242]: Please replace `var` with actual type name.

**Comment by gregsdennis** [comment]
Looking into test failures...

**Comment by gregsdennis** [comment]
Actually, I'm not sure what's going on.  I think it may have something to do with running in Linux, but I can't be sure.  Everything works in Windows and in WSL.

Windows:
image

WSL:
image

This is without further changes.

**Comment by gregsdennis** [comment]
@daxian-dbw would you mind if I put some temporary `Console.WriteLine()`s in the code to get some debugging info from the server test runner?  I'll remove them once I get it working.

**Comment by daxian-dbw** [comment]
@gregsdennis Please feel free to add debugging messages in the code or test.

**Comment by iSazonov** [comment]
@gregsdennis Perhaps it is Unix line end chars?

**Comment by gregsdennis** [comment]
No, it's having trouble finding the files (resolving `$ref`).  I'm debugging in #19617.

**Comment by gregsdennis** [comment]
Okay.  That looks good.  Now I just need to figure out this packaging stuff.

@iSazonov I think you just fixed this for me before, but I don't know what you did.  I'm happy to learn.  Is there any documentation on it?  The error just says,

> Please update D:\a\1\PowerShell\tools\packaging\Boms\windows.json per the above instructions

but there are no instructions.  I assume they're in the code somewhere?

**Comment by daxian-dbw** [comment]
Hi @gregsdennis, thanks for finding the root cause and fix the issue!
For the packaging failure, please do the following:
1. click the "View more details on Azure Pipelines" of the "PowerShell-Windows-Packaging-CI" job
2. click on the "_4 published; 1 consumed_" in the "Related" tab
3. download the `1z5wd3xl.nu4-windows-fyiwyr3s.zry-bom.json` file, and then in your working branch, replace the `PowerShell\tools\packaging\Boms\windows.json` file with it (make sure use the same `windows.json` name).
4. commit and push the change.

**Comment by gregsdennis** [comment]
Weird.  It looks like nuget is packaging my Spanish language pack now.  I'll have to look into that.

I had to update VS at some point in this process, and then I published a new version of the package.

**Comment by gregsdennis** [comment]
@daxian-dbw everything looks good from my side.  Thanks for the help.  Let me know if there's anything else you want me to change.

**Review by daxian-dbw — APPROVED** [review]
LGTM. Thanks @gregsdennis for your contribution!

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs [hunk 3]
```diff
@@ -239,32 +239,20 @@ protected override void ProcessRecord()
...
-                                WriteError(errorRecord);
-                            }
+                    EvaluationResults evaluationResults = _jschema.Evaluate(parsedJson, new EvaluationOptions { OutputFormat = OutputFormat.List });
+                    result = evaluationResults.IsValid;
+                    if (!result)
+                    {
+                        HandleValidationErrors(evaluationResults);
 
-                            if (validationResults.HasNestedResults)
+                        if (evaluationResults.HasDetails)
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs:248]: Previously, we had one top error and details errors. Now the design is changed. First of all, it is unexpected for a minor update. Secondly, it doesn't look as convenient. For example, if this cmdlet returns false, how would the user only get errors related to this check? 
@daxian-dbw Thoughts?

Comment (gregsdennis) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs:248]: No, it's the same thing. The property name just changed.  You still get all the same results. No functional change.

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs:248]: Never mind.

**Comment by iSazonov** [comment]
@gregsdennis The commit is not backported to 7.2/7.3. (And will not be). The commit will be in 7.4 Preview4.

**Comment by gregsdennis** [comment]
Thank you!

**Pull request merged**